agent/src/agent/pipeline/json_builder/json_builder.py

```diff
@@ -127,27 +133,67 @@ def _validate_config_for_create(config: dict):
     json_schema = {
         'type': 'object',
         'properties': {
-            'source': {'type': 'string', 'enum': source.repository.get_all_names()},
-            'pipeline_id': {'type': 'string', 'minLength': 1, 'maxLength': 100}
+            'source': {
+                'type': 'string',
+                'enum': source.repository.get_all_names()
+            },
+            'pipeline_id': {
+                'type': 'string',
+                'minLength': 1,
+                'maxLength': 100
+            }
         },
         'required': ['source', 'pipeline_id']
     }
     jsonschema.validate(config, json_schema)
 
 
 def _load_config(pipeline_: Pipeline, config: dict, is_edit=False):
-    if isinstance(pipeline_, pipeline.RawPipeline):
-        config['uses_schema'] = False
-    elif 'uses_schema' not in config:
-        if is_edit:
-            config['uses_schema'] = pipeline_.config.get('uses_schema', False)
-        else:
-            config['uses_schema'] = pipeline.manager.supports_schema(pipeline_)
+    config['uses_schema'] = _uses_schema(pipeline_, config, is_edit)
     json_builder.get(pipeline_, config, is_edit).build()
     # todo too many validations, 4 validations here
     pipeline.config.validators.get_config_validator(pipeline_.source.type).validate(pipeline_)
 
 
+class _SchemaChooser:
+    @staticmethod
+    def choose(pipeline_: Pipeline, config: dict, is_edit=False) -> bool:
+        if isinstance(pipeline_, pipeline.RawPipeline):
+            return False
+        elif 'uses_schema' not in config:
+            if is_edit:
+                return bool(pipeline_.config.get('uses_schema', False))
+            else:
+                return pipeline.manager.supports_schema(pipeline_)
+        else:
+            return bool(config['uses_schema'])
+
+
+class _PromQLSchemaChooser(_SchemaChooser):
+    @staticmethod
+    def choose(pipeline_: Pipeline, config: dict, is_edit=False) -> bool:
+        conf_uses = False if 'uses_schema' not in config else bool(config['uses_schema'])
+        # PromQL pipelines support schema only if dimensions and values are specified
+        actual_schema = (
+            _SchemaChooser.choose(pipeline_, config, is_edit)
+            and 'dimensions' in config and bool(config['dimensions'])
+            and 'values' in config and bool(config['values'])
+        )
+        if conf_uses and not actual_schema:
+            raise ConfigurationException(
+                'PromQL pipelines support protocol 3.0 only if `dimensions` and `values` are specified'
+            )
+        return actual_schema
+
+
+def _uses_schema(pipeline_: Pipeline, config: dict, is_edit=False) -> bool:
```

Comment:
you should have a method that returns an instance of _SchemaChooser interface with an abstract method `choose` that you will be able to call later in the _load_config method. Right now it's not clear what a `chooser` variable can be